File: android/app/src/main/kotlin/io/mynote/app/auth/AuthManager.kt
package io.mynote.app.auth

import android.content.Context
import androidx.credentials.CredentialManager
import androidx.credentials.GetCredentialRequest
import androidx.credentials.exceptions.GetCredentialException
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.firebase.FirebaseApp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

sealed interface AuthStatus {
    data object SignedOut : AuthStatus
    data class SignedIn(val uid: String, val email: String?) : AuthStatus
    /** No `google-services.json` in this build — local-only mode. */
    data object Unconfigured : AuthStatus
}

/**
 * Sign-in, backed by Firebase Auth.
 *
 * Firebase handles identity only — free for Google and email, and it takes
 * password resets and account recovery off our plate. Notes never touch
 * Firebase; they go to our own Worker.
 *
 * Signing in is always optional: MyNote is fully usable signed-out, and the
 * account exists so purchases and notes can follow the user to another device.
 */
class AuthManager(private val context: Context) {

    private val _status = MutableStateFlow<AuthStatus>(AuthStatus.SignedOut)
    val status: StateFlow<AuthStatus> = _status.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val configured: Boolean = FirebaseApp.getApps(context).isNotEmpty()

    init {
        if (!configured) {
            _status.value = AuthStatus.Unconfigured
        } else {
            Firebase.auth.addAuthStateListener { auth ->
                _status.value = auth.currentUser
                    ?.let { AuthStatus.SignedIn(it.uid, it.email) }
                    ?: AuthStatus.SignedOut
            }
        }
    }

    val isSignedIn: Boolean get() = status.value is AuthStatus.SignedIn

    /**
     * Fresh ID token for the Worker. Firebase refreshes it automatically near
     * expiry, so callers can ask on every request.
     */
    suspend fun idToken(): String? {
        if (!configured) return null
        val user = Firebase.auth.currentUser ?: return null
        return runCatching { user.getIdToken(false).await().token }.getOrNull()
    }

    suspend fun signIn(email: String, password: String) = guard {
        Firebase.auth.signInWithEmailAndPassword(email, password).await()
    }

    suspend fun signUp(email: String, password: String) = guard {
        Firebase.auth.createUserWithEmailAndPassword(email, password).await()
    }

    suspend fun sendPasswordReset(email: String) = guard {
        Firebase.auth.sendPasswordResetEmail(email).await()
    }

    /**
     * Google sign-in through Credential Manager, which is the supported path on
     * Android 14+ and degrades to Play Services on older releases.
     */
    suspend fun signInWithGoogle(activityContext: Context, serverClientId: String) = guard {
        val option = GetGoogleIdOption.Builder()
            .setServerClientId(serverClientId)
            .setFilterByAuthorizedAccounts(false)   // let them pick any account
            .build()
        val request = GetCredentialRequest.Builder().addCredentialOption(option).build()

        val result = CredentialManager.create(activityContext).getCredential(activityContext, request)
        val googleCredential = GoogleIdTokenCredential.createFrom(result.credential.data)
        val firebaseCredential = GoogleAuthProvider.getCredential(googleCredential.idToken, null)
        Firebase.auth.signInWithCredential(firebaseCredential).await()
    }

    fun signOut() {
        if (configured) Firebase.auth.signOut()
    }

    fun clearError() { _error.value = null }

    private suspend fun guard(block: suspend () -> Unit) {
        if (!configured) {
            _error.value = "Sign-in is not configured in this build."
            return
        }
        _error.value = null
        try {
            block()
        } catch (e: GetCredentialException) {
            _error.value = "Google sign-in was cancelled."
        } catch (e: Exception) {
            _error.value = friendly(e)
        }
    }

    /** Turn Firebase's exceptions into something a person can act on. */
    private fun friendly(e: Exception): String = when {
        e.message?.contains("password is invalid", true) == true ||
            e.message?.contains("INVALID_LOGIN_CREDENTIALS", true) == true ->
            "That email and password don't match."
        e.message?.contains("badly formatted", true) == true ->
            "That doesn't look like an email address."
        e.message?.contains("already in use", true) == true ->
            "There's already an account with that email. Try signing in."
        e.message?.contains("at least 6 characters", true) == true ->
            "Pick a password with at least 6 characters."
        e.message?.contains("network", true) == true ->
            "No connection. Your notes are saved on this device either way."
        e.message?.contains("blocked all requests", true) == true ->
            "Too many attempts. Wait a minute and try again."
        else -> e.message ?: "Something went wrong signing in."
    }
}

File: ios/MyNote/Sources/Auth/AuthManager.swift
import Foundation
import SwiftUI
import UIKit
import CommonCrypto
import AuthenticationServices
import FirebaseCore
import FirebaseAuth

/// Sign-in, backed by Firebase Auth.
///
/// Firebase is used for identity only — it is free for Google, Apple and email,
/// and it takes password resets, email verification and account recovery off our
/// plate entirely. Notes never touch Firebase; they go to our own Worker.
///
/// Signing in is always optional: MyNote is fully usable signed-out, and the
/// account exists so purchases and notes can follow the user to another device.
@Observable
@MainActor
final class AuthManager: NSObject {
    enum Status: Equatable {
        case signedOut
        case signedIn(uid: String, email: String?)
        /// No `GoogleService-Info.plist` in the bundle — local-only mode.
        case unconfigured
    }

    private(set) var status: Status = .signedOut
    private(set) var lastError: String?
    private(set) var isWorking = false

    private var handle: AuthStateDidChangeListenerHandle?
    private var appleNonce: String?
    private var appleContinuation: CheckedContinuation<Void, Error>?

    override init() {
        super.init()
        guard Self.isFirebaseConfigured else {
            status = .unconfigured
            return
        }
        if FirebaseApp.app() == nil { FirebaseApp.configure() }
        handle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
            Task { @MainActor in
                self?.status = user.map { .signedIn(uid: $0.uid, email: $0.email) } ?? .signedOut
            }
        }
    }

    /// The app ships without credentials so the repo stays safe to clone; a
    /// build with no plist runs local-only rather than crashing at launch.
    static var isFirebaseConfigured: Bool {
        Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil
    }

    var isSignedIn: Bool {
        if case .signedIn = status { return true }
        return false
    }

    /// Fresh ID token for the Worker. Firebase refreshes it automatically when
    /// it is close to expiry, so callers can ask on every request.
    func idToken() async throws -> String? {
        guard Self.isFirebaseConfigured, let user = Auth.auth().currentUser else { return nil }
        return try await user.getIDToken()
    }

    // MARK: - Email

    func signIn(email: String, password: String) async {
        await perform { try await Auth.auth().signIn(withEmail: email, password: password) }
    }

    func signUp(email: String, password: String) async {
        await perform { try await Auth.auth().createUser(withEmail: email, password: password) }
    }

    func sendPasswordReset(email: String) async {
        await perform { try await Auth.auth().sendPasswordReset(withEmail: email) }
    }

    // MARK: - Google

    /// Uses Firebase's web OAuth flow rather than the GoogleSignIn SDK — one
    /// fewer dependency, one fewer binary to keep current, same result.
    func signInWithGoogle() async {
        await perform {
            let provider = OAuthProvider(providerID: "google.com")
            provider.customParameters = ["prompt": "select_account"]
            let credential = try await provider.credential(with: nil)
            try await Auth.auth().signIn(with: credential)
        }
    }

    // MARK: - Apple

    /// Apple requires Sign in with Apple wherever another social login is
    /// offered, so this is a store-review requirement as much as a feature.
    func signInWithApple() async {
        let nonce = Self.randomNonce()
        appleNonce = nonce

        let request = ASAuthorizationAppleIDProvider().createRequest()
        request.requestedScopes = [.fullName, .email]
        request.nonce = Self.sha256(nonce)

        let controller = ASAuthorizationController(authorizationRequests: [request])
        controller.delegate = self
        controller.presentationContextProvider = self

        isWorking = true
        lastError = nil
        defer { isWorking = false }
        do {
            try await withCheckedThrowingContinuation { continuation in
                appleContinuation = continuation
                controller.performRequests()
            }
        } catch {
            lastError = Self.friendly(error)
        }
    }

    func signOut() {
        guard Self.isFirebaseConfigured else { return }
        try? Auth.auth().signOut()
    }

    // MARK: - Helpers

    private func perform(_ work: @escaping () async throws -> Void) async {
        guard Self.isFirebaseConfigured else {
            lastError = "Sign-in is not configured in this build."
            return
        }
        isWorking = true
        lastError = nil
        defer { isWorking = false }
        do {
            try await work()
        } catch {
            lastError = Self.friendly(error)
        }
    }

    /// Turn Firebase's error codes into something a person can act on.
    static func friendly(_ error: Error) -> String {
        let code = AuthErrorCode(rawValue: (error as NSError).code)
        switch code {
        case .wrongPassword, .invalidCredential:
            return "That email and password don't match."
        case .invalidEmail:
            return "That doesn't look like an email address."
        case .emailAlreadyInUse:
            return "There's already an account with that email. Try signing in."
        case .weakPassword:
            return "Pick a password with at least 6 characters."
        case .networkError:
            return "No connection. Your notes are saved on this device either way."
        case .tooManyRequests:
            return "Too many attempts. Wait a minute and try again."
        case .userNotFound:
            return "No account with that email yet."
        default:
            return error.localizedDescription
        }
    }

    /// A nonce binds Apple's response to this specific request, so a captured
    /// token cannot be replayed into our app.
    static func randomNonce(length: Int = 32) -> String {
        var bytes = [UInt8](repeating: 0, count: length)
        _ = SecRandomCopyBytes(kSecRandomDefault, length, &bytes)
        let charset = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-._")
        return String(bytes.map { charset[Int($0) % charset.count] })
    }

    static func sha256(_ input: String) -> String {
        var hash = [UInt8](repeating: 0, count: 32)
        let data = Array(input.utf8)
        CC_SHA256(data, CC_LONG(data.count), &hash)
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

extension AuthManager: ASAuthorizationControllerDelegate {
    nonisolated func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithAuthorization authorization: ASAuthorization
    ) {
        Task { @MainActor in
            guard
                let credential = authorization.credential as? ASAuthorizationAppleIDCredential,
                let tokenData = credential.identityToken,
                let idToken = String(data: tokenData, encoding: .utf8),
                let nonce = appleNonce
            else {
                finishApple(.failure(AuthError.appleTokenMissing))
                return
            }

            do {
                let firebaseCredential = OAuthProvider.appleCredential(
                    withIDToken: idToken,
                    rawNonce: nonce,
                    fullName: credential.fullName
                )
                try await Auth.auth().signIn(with: firebaseCredential)
                finishApple(.success(()))
            } catch {
                finishApple(.failure(error))
            }
        }
    }

    nonisolated func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithError error: Error
    ) {
        Task { @MainActor in finishApple(.failure(error)) }
    }

    private func finishApple(_ result: Result<Void, Error>) {
        let continuation = appleContinuation
        appleContinuation = nil
        appleNonce = nil
        continuation?.resume(with: result)
    }
}

extension AuthManager: ASAuthorizationControllerPresentationContextProviding {
    nonisolated func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
        MainActor.assumeIsolated {
            let scene = UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .first { $0.activationState == .foregroundActive }
            return scene?.keyWindow ?? ASPresentationAnchor()
        }
    }
}

enum AuthError: LocalizedError {
    case appleTokenMissing
    var errorDescription: String? { "Apple didn't return a sign-in token. Try again." }
}
